### PR TITLE
Fix for tearDown on this.skip(), see #2148.

### PR DESCRIPTION
**Pull request by megagon**
Tests passed, new added for checking that tearDown executed.
Something to change?

### CODE REVIEW

**Comment by megagon** [comment]
@boneskull 
Do you have any comments?

**Comment by boneskull** [comment]
see #2148 for question(s)

**Comment by laughinghan** [comment]
@megagon: this doesn't seem to do anything for the testcase in #2148, see how "Temporary stuff" still isn't cleaned up even though I modified it to reference `megagon/master`: http://jsbin.com/kubago/edit?html,output

**Comment by laughinghan** [comment]
@megagon: your PR doesn't modify `mocha.js` at all. Should it? Or is that only rebuilt on releases or something?

**Comment by ScottFreeCode** [comment]
The browser mocha.js file has to be built for release, and may be updated before then, but if I understand correct generally shouldn't be in non-blocking PRs to avoid interfering with other PRs (well, or having to rebase depending on the order of merges, anyway).

File: test/acceptance/interfaces/tdd.js [hunk 1]
```diff
@@ -38,4 +38,24 @@ suite('Array', function(){
...
+
+    teardown(function () {
+      shouldBeChangeInTearDown = 10;
+    });
+
+    suiteTeardown(function () {
+      shouldBeChangeInTearDown.should.eql(10);
+    });
+  });
 });
```
Comment (ScottFreeCode) [inline test/acceptance/interfaces/tdd.js]: I'd suggest just a boolean flag (and without duplicating "should" in its name), e.g.:
```js
var tornDown;

setup(function () {
  tornDown = false;
});

// ...

tearDown(function () {
  tornDown = true;
});

suiteTeardown(function () {
  tornDown.should.be.ok();
});
```

Comment (ScottFreeCode) [inline test/acceptance/interfaces/tdd.js]: More importantly, if this is an issue in Runner and fixed in Runner, then it shouldn't be in the TDD-specific tests, but rather in the non-interface-specific tests. (In fact, unless we've pretty severely screwed up the way the interfaces interact with the rest of the system, all we should need to test in the interface-specific tests is that all their versions of the methods work, I think...)

**Comment by ScottFreeCode** [comment]
We should determine/decide whether calling `this.skip` inside a `beforeEach`/`setup` should also still run `afterEach`/`teardown` like it will if called in the test itself, or if we should assume that `beforeEach`/`setup` is going to do any necessary cleanup if it calls `this.skip` and continue skipping `afterEach`/`teardown` in that case.

Then, we should also add a test for whichever behaviour is determined/decided to be correct in that regard.

**Comment by boneskull** [comment]
@megagon Can you please add test(s) which @ScottFreeCode would like to see to this PR?

**Comment by boneskull** [comment]
Closing as inactive; please reopen if you'd like to continue.  Thanks!

**Pull request closed without merging**